File: architecture-skills/design-patterns/examples/flyweight/flyweight.php
<?php
// Flyweight pattern in PHP.
//
// PHP copies objects by handle, not by value, so sharing is natural:
// an array-backed factory returns the same immutable (readonly)
// TreeType instance for every repeated species.

declare(strict_types=1);

/** Flyweight: immutable intrinsic state shared by many trees. */
final class TreeType
{
    public function __construct(
        public readonly string $name,
        public readonly string $color,
    ) {
    }

    public function draw(int $x, int $y): void
    {
        echo "{$this->color} {$this->name} at ($x, $y)\n";
    }
}

final class TreeTypeFactory
{
    /** @var array<string, TreeType> */
    private array $cache = [];

    public function get(string $name, string $color): TreeType
    {
        $key = "$name/$color";
        return $this->cache[$key] ??= new TreeType($name, $color);
    }

    public function count(): int
    {
        return count($this->cache);
    }
}

/** Context: extrinsic state plus a reference to the shared flyweight. */
final class Tree
{
    public function __construct(
        public readonly int $x,
        public readonly int $y,
        public readonly TreeType $kind,
    ) {
    }

    public function draw(): void
    {
        $this->kind->draw($this->x, $this->y);
    }
}

$factory = new TreeTypeFactory();
$forest = [
    new Tree(1, 1, $factory->get('oak', 'green')),
    new Tree(2, 4, $factory->get('pine', 'dark green')),
    new Tree(5, 2, $factory->get('oak', 'green')),
    new Tree(6, 6, $factory->get('birch', 'light green')),
    new Tree(8, 3, $factory->get('pine', 'dark green')),
    new Tree(9, 9, $factory->get('oak', 'green')),
];

foreach ($forest as $tree) {
    $tree->draw();
}

echo 'trees planted: ' . count($forest) . ', tree types in memory: ' . $factory->count() . "\n";

File: architecture-skills/design-patterns/examples/flyweight/flyweight.rs
// Flyweight pattern in Rust.
//
// Rc makes the sharing explicit: cloning an Rc<TreeType> copies a
// pointer, not the data, and the type system guarantees the shared
// state cannot be mutated behind other owners' backs. Use Arc instead
// of Rc when contexts cross threads.

use std::collections::HashMap;
use std::rc::Rc;

/// Flyweight: immutable intrinsic state shared by many trees.
struct TreeType {
    name: String,
    color: String,
}

impl TreeType {
    fn draw(&self, x: i32, y: i32) {
        println!("{} {} at ({}, {})", self.color, self.name, x, y);
    }
}

#[derive(Default)]
struct TreeTypeFactory {
    cache: HashMap<String, Rc<TreeType>>,
}

impl TreeTypeFactory {
    fn get(&mut self, name: &str, color: &str) -> Rc<TreeType> {
        let key = format!("{name}/{color}");
        Rc::clone(self.cache.entry(key).or_insert_with(|| {
            Rc::new(TreeType {
                name: name.to_string(),
                color: color.to_string(),
            })
        }))
    }

    fn len(&self) -> usize {
        self.cache.len()
    }
}

/// Context: extrinsic state plus a cheap handle to the shared flyweight.
struct Tree {
    x: i32,
    y: i32,
    kind: Rc<TreeType>,
}

fn main() {
    let mut factory = TreeTypeFactory::default();
    let plan = [
        (1, 1, "oak", "green"),
        (2, 4, "pine", "dark green"),
        (5, 2, "oak", "green"),
        (6, 6, "birch", "light green"),
        (8, 3, "pine", "dark green"),
        (9, 9, "oak", "green"),
    ];

    let forest: Vec<Tree> = plan
        .iter()
        .map(|&(x, y, name, color)| Tree { x, y, kind: factory.get(name, color) })
        .collect();

    for tree in &forest {
        tree.kind.draw(tree.x, tree.y);
    }

    println!("trees planted: {}, tree types in memory: {}", forest.len(), factory.len());
}
